File: backend/app/models/gae.py
from pydantic import BaseModel, Field


class GAEStep(BaseModel):
    t: int = Field(..., ge=0, description="时刻索引")
    token: str = Field(..., description="Token文本")
    r_t: float = Field(..., description="即时奖励")
    V_t: float = Field(..., description="预测价值")
    V_next: float = Field(..., description="下一时刻价值")
    delta: float = Field(..., description="TD error")
    A_t: float = Field(..., description="优势值")
    is_positive: bool = Field(..., description="优势值是否为正")


class VarianceBiasResult(BaseModel):
    bias: str = Field(..., description="偏差等级")
    variance: str = Field(..., description="方差等级")
    description: str = Field(..., description="说明文字")


class GAEResult(BaseModel):
    steps: list[GAEStep] = Field(..., description="GAE步骤列表")
    advantages: list[float] = Field(..., description="优势函数数组")
    returns: list[float] = Field(..., description="实际收益数组")
    gamma: float = Field(..., description="折扣因子")
    lambda_: float = Field(..., alias="lambda", description="GAE权衡因子")
    variance_bias: VarianceBiasResult = Field(..., description="方差-偏差解读")

    model_config = {"populate_by_name": True}


class StepCalculation(BaseModel):
    step: int = Field(..., description="当前时刻")
    formulas: dict = Field(..., description="公式计算详情")
    is_positive: bool = Field(..., description="优势值是否为正")

File: frontend/src/mock/gae.ts
import type { GAEResult, StepCalculation, GAEStep } from '@/types'
import { computeGAE, getVarianceBiasInterpretation } from '@/utils/algorithm'

export function generateGAEMockData(
  tokens: string[],
  rewards: number[],
  values: number[],
  gamma: number = 0.99,
  lambda: number = 0.95
): GAEResult {
  const { advantages, returns } = computeGAE(rewards, values, gamma, lambda)
  const varianceBias = getVarianceBiasInterpretation(lambda)

  const steps = []
  for (let t = tokens.length - 1; t >= 0; t--) {
    const vNext = t < tokens.length - 1 ? values[t + 1] : 0
    const delta = parseFloat((rewards[t] + gamma * vNext - values[t]).toFixed(4))
    steps.push({
      t,
      token: tokens[t],
      r_t: rewards[t],
      V_t: values[t],
      V_next: vNext,
      delta,
      A_t: advantages[t],
      is_positive: advantages[t] >= 0,
    })
  }

  return {
    steps,
    advantages,
    returns,
    gamma,
    lambda,
    variance_bias: varianceBias,
  }
}

export const defaultGAEMock = generateGAEMockData(
  ['Hello', 'I', 'am', 'an', 'AI', 'assistant'],
  [0.5, 0.3, 0.8, -0.1, 1.5, 0.5],
  [0.4, 0.6, 0.5, 0.3, 0.8, 0.6],
  0.99,
  0.95
)

export function generateStepCalculation(
  step: GAEStep,
  gamma: number,
  lambda: number,
  nextAdvantage: number,
): StepCalculation {
  const tdSub = `δ_${step.t} = ${step.r_t.toFixed(4)} + ${gamma.toFixed(2)}×${step.V_next.toFixed(4)} - ${step.V_t.toFixed(4)}`
  const gaeSub = `A_${step.t} = ${step.delta.toFixed(4)} + ${gamma.toFixed(2)}×${lambda.toFixed(2)}×${nextAdvantage.toFixed(4)}`

  return {
    step: step.t,
    token: step.token,
    formulas: {
      td_error: {
        formula: 'δ_t = r_t + γ·V_{t+1} - V_t',
        substitution: tdSub,
        result: step.delta,
      },
      gae: {
        formula: 'A_t = δ_t + γ·λ·A_{t+1}',
        substitution: gaeSub,
        result: step.A_t,
      },
    },
    is_positive: step.is_positive,
  }
}

export const mockStepCalculation: StepCalculation = generateStepCalculation(
  { t: 3, token: 'an', r_t: -0.1, V_t: 0.3, V_next: 0.8, delta: 0.392, A_t: 1.5205, is_positive: true },
  0.99, 0.95, 1.1999,
)
